src/features/smartSearch/components/sankeyDiagram/drawing.ts

```diff
@@ -0,0 +1,327 @@
+      this.drawInput(
+        diagCenter + totalOutWidth / 2 - mainWidth - changeWidth,
+        diagCenter + totalOutWidth / 2 - mainWidth,
+        offsetY,
+        seg.side.style
+      );
+    } else if (seg.kind == 'sub') {
+      const totalInWidth = mainWidth + changeWidth;
+
+      this.drawOutput(
+        diagCenter + totalInWidth / 2 - changeWidth,
+        diagCenter + totalInWidth / 2,
+        offsetY,
+        seg.side.style
+      );
+    }
+  }
+
+  drawEntrySegment(seg: SankeyEntrySegment, offsetY: number) {
+    const { arrowDepth, diagWidth, margin } = this.config;
+
+    const inset = this.initPathWithInset(seg.style);
+
+    const segWidth = seg.width * (diagWidth - margin * 2);
+    const diagCenter = diagWidth / 2;
+    const top = 0.3 * this.segHeight + offsetY;
+    const bottom = this.segHeight + offsetY;
+
+    this.ctx.moveTo(diagCenter - segWidth / 2 + inset, bottom);
+    this.ctx.lineTo(diagCenter - segWidth / 2 + inset, top);
+    this.ctx.lineTo(diagCenter, top + arrowDepth * 2);
+    this.ctx.lineTo(diagCenter + segWidth / 2 - inset, top);
+    this.ctx.lineTo(diagCenter + segWidth / 2 - inset, bottom);
+
+    this.ctx.stroke();
+    this.ctx.fill();
+  }
+
+  drawExitSegment(seg: SankeyExitSegment, offsetY: number) {
+    const { arrowDepth, diagWidth, margin } = this.config;
+
+    const inset = this.initPathWithInset(seg.style);
+
+    const segWidth = seg.width * (diagWidth - margin * 2);
+    const diagCenter = diagWidth / 2;
+
+    this.ctx.moveTo(diagCenter + segWidth / 2 - inset, offsetY);
+    this.ctx.lineTo(diagCenter + segWidth / 2 - inset, offsetY + arrowDepth);
+    this.ctx.lineTo(diagCenter, offsetY + arrowDepth * 3);
+    this.ctx.lineTo(diagCenter - segWidth / 2 + inset, offsetY + arrowDepth);
+    this.ctx.lineTo(diagCenter - segWidth / 2 + inset, offsetY);
+
+    this.ctx.stroke();
+    this.ctx.fill();
+  }
+
+  drawInput(
+    bottomLeftX: number,
+    bottomRightX: number,
+    offsetY: number,
+    style: SEGMENT_STYLE
+  ) {
+    const { arrowDepth, arrowWidth } = this.config;
+    const segHeight = this.segHeight;
+
+    const inset = this.initPathWithInset(style);
+
+    this.ctx.moveTo(0, segHeight / 2 + arrowWidth / 2 + offsetY);
+    this.ctx.lineTo(arrowDepth, segHeight / 2 + offsetY);
+    this.ctx.lineTo(0, segHeight / 2 - arrowWidth / 2 + offsetY);
+    this.ctx.quadraticCurveTo(
+      bottomRightX - inset,
+      segHeight / 2 - arrowWidth / 2 + offsetY,
+      bottomRightX - inset,
+      segHeight + offsetY
+    );
+    this.styledLineTo(bottomLeftX + inset, segHeight + offsetY, style);
+    this.ctx.quadraticCurveTo(
+      bottomLeftX + inset,
+      segHeight / 2 + arrowWidth / 2 + offsetY,
+      0,
+      segHeight / 2 + arrowWidth / 2 + offsetY
+    );
+
+    this.ctx.stroke();
+    this.ctx.fill();
+  }
+
+  drawMain(
+    upperLeftX: number,
+    upperRightX: number,
+    lowerRightX: number,
+    lowerLeftX: number,
+    offsetY: number,
+    style: SEGMENT_STYLE
```

Comment:
love that these are fully spelled out. even if i don't understand what's going on lol i appreciate it :)

Reply:
I thought about you every step of the way! 😊